### ISSUE TITLE
airmon-ng: invalid rfkill command against soft block

### ISSUE THREAD

**Issue by aircrack-ng** [aircrack-ng/aircrack-ng#1723]
**Reported by kcdtv on 12 Jun 2017 14:56 UTC**

**Context**: 
If airmon-ng detect a soft block that impeaches to activate monitor mode user is asked is he wants to auomaticaly remove the soft block

```
root@kalimuX0:~# rfkill block 1
root@kalimuX0:~# airmon-ng start wlan0

Found 2 processes that could cause trouble.
If airodump-ng, aireplay-ng or airtun-ng stops working after
a short period of time, you may want to run 'airmon-ng check kill'

  PID Name
  622 wpa_supplicant
 1596 NetworkManager

PHY	Interface	Driver		Chipset

phy0	wlan0		ath9k		Qualcomm Atheros QCA9565 / AR9565 Wireless Network Adapter (rev 01)
	wlan0 is soft blocked, please run "rfkill unblock wlan0" to use this interface.
rfkill error, unable to start wlan0

Would you like to try and automatically resolve this? [

```

**Issue**:
 - Command proposed is wrong (it should be index instead of wlan) 
 - Executed command fails (wrong too)
```
Would you like to try and automatically resolve this? [y/n](y/n]) y

Failed to set wlan0mon up using ip

		(mac80211 monitor mode vif enabled for [on [phy0](phy0]wlan0)wlan0mon)
		(mac80211 station mode vif disabled for [Semiconductor Corp. 
```
So By answering "yes" an incorrect command is executed and soft block remains

**Primitive walk around to get correct behaviour**:
I found 2 lines that , i guess, are part of the problem:
  1) Declaration of the variable **$index** in line **268** doesn't look correct to me
Replacing line 268 by this one
```
index=$(rfkill list | grep ${1} | awk -F ':' '{print $1}')
```
Does the job to have a correct index number 
 2) second point replace the array by the variable $index in the **rfkill command** (line **305**) 
like this
```
rfkill_status=$(rfkill unblock $index 2>&1)
```
  Now the soft block is correctly removed:

```
root@kalimuX0:~# rfkill list
0: Toshiba Bluetooth: Bluetooth
	Soft blocked: no
	Hard blocked: no
1: phy0: Wireless LAN
	Soft blocked: yes
	Hard blocked: no
3: hci0: Bluetooth
	Soft blocked: no
	Hard blocked: no
5: phy3: Wireless LAN
	Soft blocked: no
	Hard blocked: no
root@kalimuX0:~# airmon-ng start wlan0

Found 2 processes that could cause trouble.
If airodump-ng, aireplay-ng or airtun-ng stops working after
a short period of time, you may want to run 'airmon-ng check kill'

  PID Name
  622 wpa_supplicant
 1596 NetworkManager

PHY	Interface	Driver		Chipset

phy0	wlan0		ath9k		Qualcomm Atheros QCA9565 / AR9565 Wireless Network Adapter (rev 01)
: ------DEBUG-----
index varibale value: 
    -  1
-----END DEBUG-----
	wlan0 is soft blocked, please run "rfkill unblock wlan0" to use this interface.
rfkill error, unable to start wlan0

Would you like to try and automatically resolve this? [y/n](phy0]wlan0)
phy3	wlan1		8814au		Realtek) y
: ------DEBUG-----
executed command: 
    -  
-----END DEBUG-----

		(mac80211 monitor mode vif enabled for [on [phy0](phy0]wlan0)wlan0mon)
		(mac80211 station mode vif disabled for [phy0]wlan0)
phy3	wlan1		8814au		Realtek Semiconductor Corp. 

root@kalimuX0:~# rfkill list
0: Toshiba Bluetooth: Bluetooth
	Soft blocked: no
	Hard blocked: no
1: phy0: Wireless LAN
	Soft blocked: no
	Hard blocked: no
3: hci0: Bluetooth
	Soft blocked: no
	Hard blocked: no
5: phy3: Wireless LAN
	Soft blocked: no
	Hard blocked: no

```
 
Cheers

**Comment by aircrack-ng** [comment]
**Comment by kcdtv on 12 Jun 2017 14:56 UTC**

rectification:
After checking: It has nothing to do with the way index is declared. Just the second mod (use index instead of phy array ) is enough

**Comment by aircrack-ng** [comment]
**Comment by misterx on 12 Jun 2017 14:56 UTC**

Could you post a patch? Basically, get the sources from subversion, do your modifications and then run ```svn diff > airmon-ng_rfkill_fix_1723.diff``` and upload it here? It saves you and me a lot of time (you don't have to explain the modifications (because the lines you mention don't match with current code; patch can handle a bit of change in the code).

**Comment by aircrack-ng** [comment]
**Attachment by kcdtv on 15 Jun 2017 12:19 UTC**

Uploaded file [airmon-ng_rfkill_fix_1723.diff](https://github.com/aircrack-ng/trac-attachments/blob/master/1723/20170615-121913_246-airmon-ng_rfkill_fix_1723.diff)

"rfkill unblock" command corrected

**Comment by aircrack-ng** [comment]
**Comment by misterx on 12 Jun 2017 14:56 UTC**

Thanks.

Do you have any idea when it changed? I'm pretty sure the code used to work when it was committed but something changed in rfkill along the route. Or is it backward-compatible?

**Comment by aircrack-ng** [comment]
**Comment by kcdtv on 12 Jun 2017 14:56 UTC**


```
Do you have any idea when it changed?
```

Sorry: I was pretty busy and didn't see your question. 
The error appears somewhere between **aircrack-ng-1.2-rc3** (it works correctly) and **aircrack-ng-1.2-rc4** (It doesn't work correctly) 
Cheers

**Comment by aircrack-ng** [comment]
**Comment by misterx on 12 Jun 2017 14:56 UTC**

I can't find anything related to rfkill between rc3 and rc4.

**Comment by aircrack-ng** [comment]
**Comment by misterx on 12 Jun 2017 14:56 UTC**

Should be fixed in aircrack-ng/aircrack-ng@a583f5b5f23f86ee2403eef01de986be46038070, it will try both commands. Please test and reopen if it doesn't fix it.

**Comment by aircrack-ng** [comment]
**Modified by misterx on 12 Jun 2017 14:56 UTC**
